src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -0,0 +1,294 @@
+      original : createRecordWithId(original);
+    return new BSONWritable(extractRecord(record, DEFAULT_ID_FIELD_NAME));
+  }
+
+  private StructuredRecord createRecordWithId(StructuredRecord original) {
+    Schema.Field idField = original.getSchema().getField(idFieldName);
+
+    if (idField == null) {
+      throw new UnexpectedFormatException(String.format("Record does not contain identifier field '%s'.", idFieldName));
+    }
+
+    Schema.Field defaultIdField = original.getSchema().getField(DEFAULT_ID_FIELD_NAME);
+    if (defaultIdField != null) {
+      throw new UnexpectedFormatException(String.format("Record already contains identifier field '%s'.",
+                                                        DEFAULT_ID_FIELD_NAME));
+    }
+
+    List<Schema.Field> copiedFields = Objects.requireNonNull(original.getSchema().getFields()).stream()
+      .filter(field -> !idFieldName.equals(field.getName()))
+      .collect(Collectors.toList());
+
+    List<Schema.Field> recordWithIdFields = new ArrayList<>();
+    recordWithIdFields.add(Schema.Field.of(DEFAULT_ID_FIELD_NAME, idField.getSchema()));
+    recordWithIdFields.addAll(copiedFields);
+
+    String recordName = Objects.requireNonNull(original.getSchema().getRecordName(), "Record name can not be empty");
+    Schema recordWithIdSchema = Schema.recordOf(recordName, recordWithIdFields);
+    StructuredRecord.Builder builder = StructuredRecord.builder(recordWithIdSchema);
+    for (Schema.Field field : original.getSchema().getFields()) {
+      if (idFieldName.equals(field.getName())) {
+        builder.set(DEFAULT_ID_FIELD_NAME, original.get(idFieldName));
+        continue;
+      }
+
+      builder.set(field.getName(), original.get(field.getName()));
+    }
+
+    return builder.build();
+  }
+
+
+  private DBObject extractRecord(Object value, @Nullable String idFieldName) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+    StructuredRecord record = (StructuredRecord) value;
+    List<Schema.Field> fields = Objects.requireNonNull(record.getSchema().getFields(), "Schema fields cannot be empty");
+    BasicDBObjectBuilder bsonBuilder = BasicDBObjectBuilder.start();
+    for (Schema.Field field : fields) {
+      Schema.LogicalType fieldLogicalType = field.getSchema().isNullable()
+        ? field.getSchema().getNonNullable().getLogicalType()
+        : field.getSchema().getLogicalType();
+
+      if (!Strings.isNullOrEmpty(idFieldName) && idFieldName.equals(field.getName())) {
+        bsonBuilder.add(idFieldName, extractObjectId(record.get(field.getName())));
+        continue;
+      }
+
+      if (fieldLogicalType == null) {
+        bsonBuilder.add(field.getName(), extractValue(record.get(field.getName()), field.getSchema()));
+        continue;
+      }
+
+      // Get values of logical types properly
+      switch (fieldLogicalType) {
```

Comment:
this logic should be moved to extractValue(). Otherwise this won't be called when logical types are used within a map or array.

Reply:
[Moved.](https://github.com/data-integrations/mongodb-plugins/pull/4/files#diff-193a88503f826159a20dd58ddbefedb8R132)

[testTransformNestedLogicalTypes](https://github.com/data-integrations/mongodb-plugins/pull/4/files#diff-ecf028edd318a21770bdeca8164196eaR538) test created.